### ISSUE TITLE
False positive of `no-import-module-exports` on `module` variable defined in the module

### ISSUE THREAD

**Issue by zloirock** [import-js/eslint-plugin-import#2181]
```js
import 'something'; // => Cannot use import declarations in modules that export using CommonJS (module.exports = 'foo' or exports.bar = 'hi')  import/no-import-module-exports

export const module = {};
module.foo = 42;
```

**Comment by ljharb** [comment]
Is the variable “module” allowed on strict mode?

**Comment by zloirock** [comment]
[Sure, it's allowed.](https://tc39.es/ecma262/#sec-strict-mode-of-ecmascript)

**Comment by ljharb** [comment]
oh right, duh. thanks.

**Comment by micalevisk** [comment]
I'm facing this in `eslint-plugin-import@2.25.3` after upgrading `eslint-config-airbnb-base` to 15.0.0

**Comment by ljharb** [comment]
Per #2297, this should be fixed in v2.25.3.

@micalevisk can you share the code that’s erroring?

**Comment by micalevisk** [comment]
This is a TypeScript project that outputs Commonjs, btw.

every line of this:
![image](https://user-images.githubusercontent.com/13461315/141164556-3d09481e-8c5e-4530-b39d-f5f1d99c9dd4.png)
![image](https://user-images.githubusercontent.com/13461315/141164613-991bf8cb-614d-4543-ad14-18287779a8cd.png)

and this:

![image](https://user-images.githubusercontent.com/13461315/141164930-22a455b3-92d9-47ab-a479-0e34c10c1722.png)
![image](https://user-images.githubusercontent.com/13461315/141164970-bab8b37b-00d4-4aeb-b294-16a2e6cad56b.png)

<details>
<summary><code>.eslintrc.js</code></summary>

```js
module.exports = {
  root: true,

  reportUnusedDisableDirectives: true,

  parserOptions: {
    ecmaVersion: 2020,
  },

  env: {
    es6: true,
    node: true,
    jest: true,
  },

  globals: {},

  extends: ['plugin:prettier/recommended'],
  plugins: ['import-helpers'],
  // Global rules
  rules: {
    'no-console': 'off',
    'no-bitwise': 'off',
    'no-plusplus': ['error', { allowForLoopAfterthoughts: true }],

    'max-len': 'off',

    'import-helpers/order-imports': [
      'error',
      {
        newlinesBetween: 'always',
        groups: [['module', '/^@[^/]/'], '/^@//', ['parent', 'sibling', 'index']],
        alphabetize: {
          order: 'asc',
          ignoreCase: true,
        },
      },
    ],
    'import/no-extraneous-dependencies': [
      'error',
      {
        packageDir: '.',
        // Files with scripts/modules that will be use only in development
        devDependencies: ['test/*', 'jest*', 'configs/**/*', 'tools/**/*'],
      },
    ],
  },

  overrides: [
    /* ==================== Dealing with TypeScript files ==================== */
    {
      files: ['*.ts'],
      parser: '@typescript-eslint/parser',
      parserOptions: {
        project: './tsconfig.lint.json',
        sourceType: 'module',
        ecmaVersion: 2020,
      },

      settings: {
        'import/parsers': {
          '@typescript-eslint/parser': ['.ts'],
        },
        'import/resolver': {
          typescript: {
            project: './tsconfig.lint.json',
          },
        },
      },

      plugins: ['jest', '@typescript-eslint', 'import'],
      extends: [
        'airbnb-base',
        'plugin:@typescript-eslint/recommended',
        'plugin:import/typescript',
        'plugin:@typescript-eslint/recommended-requiring-type-checking',
        'plugin:prettier/recommended',
        'plugin:jest/recommended',
      ],
      rules: {
        'jest/expect-expect': [
          'error',
          {
            assertFunctionNames: ['expect', 'request.**.expect'],
          },
        ],

        'no-shadow': 'off', // Disable this one in favor of the TS-version one
        '@typescript-eslint/no-shadow': 'error',

        'default-param-last': 'off', // Disable this one in favor of the TS-version one
        '@typescript-eslint/default-param-last': 'error',

        '@typescript-eslint/no-unused-vars': [
          'warn',
          { ignoreRestSiblings: true, argsIgnorePattern: '^_' },
        ],
        '@typescript-eslint/no-useless-constructor': 'error',
        '@typescript-eslint/no-explicit-any': 'off',
        // '@typescript-eslint/no-unsafe-return': 'off',
        '@typescript-eslint/no-non-null-assertion': 'off',
        '@typescript-eslint/explicit-module-boundary-types': [
          'error',
          {
            allowArgumentsExplicitlyTypedAsAny: true,
            allowDirectConstAssertionInArrowFunctions: true,
            allowedNames: [],
            allowHigherOrderFunctions: true,
            allowTypedFunctionExpressions: true,
          },
        ],

        'import/order': 'off',
        'import/prefer-default-export': 'off',
        'import/no-unresolved': 'off', // Let TS handle this
        'import/named': 'off', // and this, due to some resolution error
        'import/extensions': [
          'error',
          {
            json: 'always',
            ts: 'never',
            tsx: 'never',
            js: 'never',
            jsx: 'never',
          },
        ],
        'import/no-extraneous-dependencies': [
          'off', // Let typescript compiler complains about this.
          {
            // In case we want to enable this rule.
            packageDir: '.',
            // Files with scripts/modules that will be use only in development
            devDependencies: ['**/*spec.ts', '**/global.d.ts', '**/*.seed.ts', '**/*.factory.ts'],
          },
        ],

        /* ==================== overrides Airbnb's rules ==================== */
        'class-methods-use-this': 'off',
        'max-classes-per-file': ['error', 2], // Max 2 classes per file
        'max-len': [
          'error',
          {
            code: 100, // same as `printWidth` from Prettier
            comments: 83,
            ignoreTrailingComments: true,
            ignoreUrls: true,
            ignoreStrings: true,
            ignoreTemplateLiterals: true,
            ignoreRegExpLiterals: true,
          },
        ],
        'no-bitwise': 'off',
        'no-useless-constructor': 'off',
        'no-use-before-define': 'off',
        'no-unused-vars': 'off',
        'no-underscore-dangle': 'off',
        // Overrides https://github.com/airbnb/javascript/blob/b6fc6dc7c3cb76497db0bb81edaa54d8f3427796/packages/eslint-config-airbnb-base/rules/style.js#L257
        // See https://github.com/airbnb/javascript/issues/1271
        'no-restricted-syntax': ['error', 'ForInStatement', 'LabeledStatement', 'WithStatement'],
      },
    },

    /* ======================= Dealing with edge cases ======================= */
    {
      // Scripts-like files under `src` directory.
      files: ['src/scripts/**/*'],
      rules: {
        'no-console': 'off',
      },
    },
    {
      // Entities/models typescript files under `src` directory.
      files: ['src/**/*.entity.ts'],
      rules: {
        // Allow define relations between database tables, while using JS classes.
        // NOTE: This leave rooms to `undefined` values when using `import` in these
        //       modules.
        'import/no-cycle': 'off',
      },
    },

    /* ======================= Dealing with Jest files ======================= */
    {
      files: ['**/*spec.ts'],
      rules: {
        'max-classes-per-file': 'off',
        'no-multi-assign': 'off',

        'import/no-extraneous-dependencies': 'off',

        '@typescript-eslint/no-empty-function': 'off',
        '@typescript-eslint/explicit-module-boundary-types': 'warn',
        '@typescript-eslint/unbound-method': 'off',
      },
    },
  ],
};
```

</details>

---

```
$  grep version node_modules/eslint-plugin-import/package.json 
  "version": "2.25.3",

$  grep version node_modules/eslint-config-airbnb-base/package.json 
  "version": "15.0.0",
```

**Comment by ljharb** [comment]
You need to set `sourceType: 'module'` in parserOptions if you're using ESM syntax. Does that fix it?

**Comment by micalevisk** [comment]
> You need to set sourceType: 'module' in parserOptions

I'm using it already :/

![image](https://user-images.githubusercontent.com/13461315/141195474-26630de0-e124-4a9e-b60d-ac590c05a35d.png)

this error arises when I use `eslint-config-airbnb-base@15.0.0` (`eslint` v8.2.0), with v14 it worked. I didn't investigate this further yet

**Comment by ljharb** [comment]
@micalevisk if you could set up a repro repo that would be most helpful.

**Comment by micalevisk** [comment]
I figured out that I'm getting those lint errors when I access any property of [`module` variable](https://nodejs.org/api/modules.html#module) that was created by webpack. 

![image](https://user-images.githubusercontent.com/13461315/141651206-cc0b5417-e8cc-4109-a923-e6c340e4de33.png)

![image](https://user-images.githubusercontent.com/13461315/141651225-39336286-fc02-4022-abcb-58aeb946589c.png)

Here's the code: https://gitlab.com/micalevisk/eslint-plugin-import-issue-2181

I might be missing something. I appreciate your help @ljharb

**Comment by ljharb** [comment]
`module` is something that's only supposed to exist in CJS - if webpack is creating it in ESM, that's going to break lots of static analysis.

**Comment by micalevisk** [comment]
this is a TS project that targets CJS, not ESM. Maybe I need to configure some plugin then?

**Comment by ljharb** [comment]
What's targeted shouldn't matter, only the source; `module` isn't available in ESM, and your source is ESM.

My guess is that there's a way webpack can use `import.meta` instead of `module`, and you'd both have less confusing code, and also bypass this overly-aggressive lint rule (that said, i'm still happy to remove some false positives)

**Comment by micalevisk** [comment]
right

unfortunately I can't use `import` because the [module system](https://www.typescriptlang.org/tsconfig/#module) I'm using is CJS (due to `"module": "commonjs"` entry on my tsconfig)

```
The 'import.meta' meta-property is only allowed when the '--module' option is 'es2020', 'esnext', or 'system'.ts(1343)
```

**Comment by snitin315** [comment]
@ljharb I can reproduce it with the following code as per `eslint-plugin-import@2.26.0`:

```
import React from 'react'; // error   Cannot use import declarations in modules that export using ...
import { render } from 'react-dom'; // error
import { loadableReady } from '@loadable/component'; // error
import { BrowserRouter } from 'react-router-dom'; // error
import { HelmetProvider } from 'react-helmet-async'; // error


if (module.hot) {
  render(app, root);
  module.hot.accept();
} else {
  loadableReady(() => {
    hydrateApp(app, root);
  });
}
```

If I remove the `module.hot` condition there is no error, the following code is lint free.

```js
import React from 'react';
import { render } from 'react-dom'; 
import { loadableReady } from '@loadable/component'; 
import { BrowserRouter } from 'react-router-dom';
import { HelmetProvider } from 'react-helmet-async';
```

**Comment by mmeqbel-wiley** [comment]
> What's targeted shouldn't matter, only the source; `module` isn't available in ESM, and your source is ESM.
> 
> My guess is that there's a way webpack can use `import.meta` instead of `module`, and you'd both have less confusing code, and also bypass this overly-aggressive lint rule (that said, i'm still happy to remove some false positives)

webpack exposes the hot-module-replacement interface also under import.meta.webpackHot that can be used in [strict ESM]. but that doesn't solve the issue for me

**Comment by ljharb** [comment]
@snitin315 i definitely can reproduce the warnings, but they seem valid to me - `module` isn't in scope in ES Modules. I'm not sure how to resolv

**Comment by sangaline** [comment]
@snitin315 The solution for the `module.hot` issue with webpack's [`HotModuleReplacementPlugin](https://webpack.js.org/plugins/hot-module-replacement-plugin`) is to use [`import.meta.webpackHot`](https://webpack.js.org/api/module-variables/#importmetawebpackhot).

**Comment by ljharb** [comment]
actually reopening for the false positives; but “webpack” isn’t a problem per the above.
